opennms-doc/guide-development/src/asciidoc/text/graph/graph.adoc

```diff
@@ -0,0 +1,201 @@
+
+
+[[ga-development-graph-service-api-terminology]]
+=== Terminology
+
+A *Graph* consists of any number of points with any number of connections in between.
+Usually these points are called *Nodes*, but as *Nodes* have a specific meaning in _{opennms-product-name}_ in the context of the _Graph Service_ they are called *Vertices* (plural of *Vertex*).
+The connection between two *Vertices* are usually called *Link*, however it also has a specific meaning inside _{opennms-product-name}_, thus it is called *Edge*.
+A *Graph Container* contains any number of *Graphs*.
+In general an *Element* can either be a *Graph*, *Vertex* or *Edge*.
+
+
+=== Graph Model
+
+[[ga-development-graph-service-api-generic-model]]
+==== Generic Graph Model
+
+The _Generic Graph Model_ is very similar to link:http://graphml.graphdrawing.org[GraphML]'s model and consists of the basic elements:
+
+* `GenericGraphContainer`,
+* `GenericGraph`,
+* `GenericVertex` and
+* `GenericEdge`
+
+Each of the above have besides their designated functionality described in link:#ga-development-graph-service-api-terminology[Terminology] also properties assigned to them.
+Some properties are mandatory and determined by the element's type.
+Others are optionally and totally up to the entity providing a graph to the system.
+
+The following rules apply to the _Generic Graph Model_:
+
+- Each Graph container must be uniquely identified.
+  This identifier is called a `container id`
+- Each Graph must be uniquely identified overall graphs (meaning the namespace must be unique through the system).
+  This identifier is called a `namespace`.
+- Each Vertex and Edge must be uniquely identified by an identifier.
+  This identifier is called an `id`.
+- Each Vertex and Edge have the same `namespace` as the Graph they are part of.
+  This allows to uniquely identify vertices and edges throughout the system.
+- Each Graph, Vertex or Edge have additional properties to define their nature in more detail.
+  E.g. a Vertex may contain a `label`, `node` or `location` property.
+- Each Edge contains the source and target id of the Vertex it connects, therefore is ALWAYS directed.
+- Each edge may reference Vertices from a different Graph (`namespace`).
+  However at least one side must share the same `namespace` as the edge itself.
+  This is the "owning" side, thus the edge will be a member of that Graph.
+
+===== Identifier and Namespace
+
+In order to uniquely identify each element and each graph container within the system different strategies are applied.
+
+1. The container is uniquely identified by an `id` property.
+2. The graph container is uniquely identified by a `namespace` property.
+3. Each element within the graph is uniquely identified by a `namespace` and `id` property.
+The `id` property must not be unique in the system, but within the graph.
+
+If a graph defines an `id` property as well, this SHOULD but MUST not be identically to the `namespace` property.
```

Comment:
Reading this now I wonder why we do that? That seems to make things more complicated than they have to. Shouldn't we merge the id and the namespace and use only one?

Reply:
The `id` property is only there to be compatible with GraphML.
This means, you can set an id property on the graph, but it is identified using the namespace. In order to make things easier, simply make them the same. 
This is more or less only relevant when importing graphs via GraphML.

Reply:
hm ok, I think you are right. It is the same thing but since the children of a graph (Vertex and Edge) need to reference the graph but have their own attribute named "id" we had to come up with a different name. In the end we are talking about the graph identifier. Let's leave it as is.

Reply:
As mentioned above the separate `Namespace and Identifier` chapter is not ideal. When rearranging (or consolidating) things I will consider rewriting this a bit to make it more clear.